### PR TITLE
[ADD] website_sale_slides: add default course product

### PR DESCRIPTION
**Pull request by groundziro**
Purpose
=======
This commit adds a default product called "Access Course" to the
database when you have the `website_sale_slides` installed.

This commit also starts the transition for courses products
into `event ticket`-like products

Specifications
==============
This product is already created in order to help the users start to sell
their courses.

We also added a new type of product named `Courses`.
This will be used to create courses and easily work with courses products.

The commit adds the list of all the courses that the will gain access to
when he will buy the product inside of the cart and in the right column
whe it appears.

It also adds the same list when the product is inserted into a sales order.

The commit adds the choice of enroll policies in the front-end when
`website_sales_slides` is installed and of the corresponding product
if we are creating a paid course.

task-2636092

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by std-odoo — COMMENTED** [review]
Hi :) 

First pass of validation, Thanks for your work :)

**Review by std-odoo — COMMENTED** [review]
Re :) 

Just a few more comments and we are good :)

**Review by std-odoo — COMMENTED** [review]
@thjo-odoo Hi :) I added a few comment, we are almost good :)

**Review by std-odoo — COMMENTED** [review]
just 3 comments left, then I green the task :)

File: addons/website_sale_slides/models/sale_order.py [hunk 1]
```diff
@@ -17,15 +17,17 @@ def _action_confirm(self):
...
         )
         channel_products = related_channels.mapped('product_id')
 
         channels_per_so = {sale_order: self.env['slide.channel'] for sale_order in self}
         for so_line in so_lines:
             if so_line.product_id in channel_products:
+                so_line.name = 'Access to:'
                 for related_channel in related_channels:
                     if related_channel.product_id == so_line.product_id:
+                        so_line.name += '\n%s' % related_channel.name
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/models/sale_order.py:30]: Shouldn't we instead rely on ``get_sale_order_line_multiline_description_sale`` ? I don't think purpose of confirm action is to update line names.

File: addons/website_sale_slides/models/slide_channel.py [hunk 2]
```diff
@@ -70,3 +71,24 @@ def _filter_add_members(self, target_partners, **member_values):
             else:
                 result |= on_payment
         return result
+
+    def default_get(self, default_fields):
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/models/slide_channel.py:75]: Should be defined before the fields, as per guidelines.

File: addons/website_sale_slides/models/slide_channel.py [hunk 3]
```diff
@@ -70,3 +71,24 @@ def _filter_add_members(self, target_partners, **member_values):
...
+        values = super(Channel, self).default_get(default_fields)
+
+        if 'product_id' in default_fields:
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/models/slide_channel.py:78]: Why not a real default on product_id field ?

File: addons/website_sale_slides/models/slide_channel.py [hunk 4]
```diff
@@ -70,3 +71,24 @@ def _filter_add_members(self, target_partners, **member_values):
...
+            product_courses = self.env['product.product'].search(
+                [('detailed_type', '=', 'course')], limit=2)
+            if len(product_courses) == 1:
+                values['product_id'] = product_courses.id
+
+        return values
+
+    @api.model
+    def get_course_products(self):
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/models/slide_channel.py:87]: I would put it into a controller and behind a (technical) route as basically it has nothing to do with channel model.

File: addons/website_sale_slides/models/product_product.py [hunk 5]
```diff
@@ -8,3 +8,11 @@ class Product(models.Model):
...
 
     channel_ids = fields.One2many('slide.channel', 'product_id', string='Courses')
+
+    def get_product_multiline_description_sale(self):
+        payment_channels = self.channel_ids.filtered(lambda course: course.enroll == 'payment')
+
+        if not payment_channels:
+            return super(Product, self).get_product_multiline_description_sale()
+
+        return 'Access to:\n' + '\n'.join(payment_channels.mapped('name'))
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/models/product_product.py:18]: Translation marks ! And maybe use a better formating (like ``_('Access to %(formatted_channel_names)s', formatted_channel_name= ...)``).

File: addons/website_sale_slides/models/sale_order.py [hunk 6]
```diff
@@ -17,15 +17,17 @@ def _action_confirm(self):
...
         related_channels = self.env['slide.channel'].search(
-            [('product_id', 'in', products.ids)]
+            [('product_id', 'in', products.ids), ('enroll', '=', 'payment')],
         )
         channel_products = related_channels.mapped('product_id')
 
         channels_per_so = {sale_order: self.env['slide.channel'] for sale_order in self}
         for so_line in so_lines:
             if so_line.product_id in channel_products:
+                so_line.name = 'Access to:'
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/models/sale_order.py:27]: Missing translation marks (+ would be better to create it using a string formatted to ease translations)

File: addons/website_sale_slides/views/website_slides_templates.xml [hunk 7]
```diff
@@ -139,4 +139,14 @@
     </xpath>
 </template>
 
+<template id="cart_summary_inherit_website_sale_slides"
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/views/website_slides_templates.xml:142]: After merge of https://github.com/odoo/odoo/pull/77050 would be moved into a website_sale_templates.xml file (added in that PR :) )

File: addons/website_slides/static/src/scss/website_slides.scss [hunk 8]
```diff
@@ -573,6 +573,11 @@ $o-wslides-fs-side-width: 300px;
     box-shadow: 3px 3px 5px gray;
 }
 
+#slide_channel_add_form label img {
```
Comment (tde-banana-odoo) [inline addons/website_slides/static/src/scss/website_slides.scss:576]: Mmh ? At least provide an explanation maybe ?

**Review by tde-banana-odoo — COMMENTED** [review]
Review ! Some comments (try to be closer of standard sale-like behavior), some missing stuff (translation marks). Cheers !

File: addons/website_sale_slides/models/sale_order.py (hunk 1 above)
Comment (std-odoo) [reply addons/website_sale_slides/models/sale_order.py:30]: Indeed, old code I forgot to remove :/

File: addons/website_sale_slides/models/sale_order.py (hunk 6 above)
Comment (std-odoo) [reply addons/website_sale_slides/models/sale_order.py:27]: Removed (see above)

File: addons/website_sale_slides/views/website_slides_templates.xml (hunk 7 above)
Comment (std-odoo) [reply addons/website_sale_slides/views/website_slides_templates.xml:142]: Not sure it's the same template, nut I moved it in `website_sale_template` as indeed it makes more sense

File: addons/website_sale_slides/controllers/__init__.py [hunk 9]
```diff
@@ -1,4 +1,5 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+from . import main
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/controllers/__init__.py:4]: We already have an override of slides controllers just below ^^ (I think it came with the rebase on top of the other slides promenade)

File: addons/website_sale_slides/controllers/main.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+        """Return a list of the course products values with formatted price."""
+        products = request.env['product.product'].search([('detailed_type', '=', 'course')])
+
+        return [{
+            'id': product.id,
+            'name': f'{product.name} ({format_amount(request.env, product.list_price, product.currency_id)})',
+        } for product in products]
+
+
+    def _slide_channel_prepare_values(self, **kwargs):
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/controllers/main.py:23]: Maybe add a docstring to explain it ? It is not crystal clear the purpose of this override when reading it without knowing a bit how it works under the radar.

File: addons/website_sale_slides/models/slide_channel.py [hunk 11]
```diff
@@ -7,10 +7,16 @@
...
+        product_courses = self.env['product.product'].search(
+            [('detailed_type', '=', 'course')], limit=2)
+        return product_courses.id if len(product_courses) == 1 else False
+
     enroll = fields.Selection(selection_add=[
         ('payment', 'On payment')
     ], ondelete={'payment': lambda recs: recs.write({'enroll': 'invite'})})
-    product_id = fields.Many2one('product.product', 'Product', index=True)
+    product_id = fields.Many2one('product.product', 'Product', domain=[('detailed_type', '=', 'course')],
+                                 index=True, default=_get_default_product_id)
```
Comment (tde-banana-odoo) [inline addons/website_sale_slides/models/slide_channel.py:19]: While we are at it, do we really need an index ? Seems channel is not the most heavily used table ...

**Review by tde-banana-odoo — COMMENTED** [review]
Ririviou ! Looks good, some side comments :)

File: addons/website_sale_slides/controllers/__init__.py (hunk 9 above)
Comment (std-odoo) [reply addons/website_sale_slides/controllers/__init__.py:4]: ah maybe, didn't notice, good catch :)

**Review by std-odoo — COMMENTED** [review]
I've made the changes :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Pull request closed without merging**